mlx/array.cpp

```diff
@@ -167,6 +166,32 @@ void array::move_shared_buffer(array other) {
   move_shared_buffer(other, other.strides(), other.flags(), other.data_size());
 }
 
+array::~array() {
+  if (array_desc_ == nullptr) {
+    return;
+  }
+
+  // Break circular reference for non-detached arrays with siblings
+  if (auto n = siblings().size(); n > 0) {
+    bool do_detach = true;
+    do_detach &= (array_desc_.use_count() == (n + 1));
+    for (auto& s : siblings()) {
+      do_detach &= (s.array_desc_.use_count() == n);
+      if (!do_detach) {
+        break;
+      }
+    }
+    if (do_detach) {
+      for (auto& s : siblings()) {
+        for (auto& ss : s.siblings()) {
+          ss.array_desc_ = nullptr;
+        }
+        s.array_desc_->siblings.clear();
```

Comment:
I may be particularly slow today but I think the above code has a bug. The siblings of siblings will set the `array_desc_` to `nullptr` and then we 'll try to dereference that to clear the siblings.

```
Let our arrays be A1, A2, A3 all being siblings and the corresponding siblings vectors being S1, S2, S3.

S1 = [A2, A3]
S2 = [A1, A3]
S3 = [A1, A2]

Suppose A1 is the array being destroyed and A2, A3 have no external references the following lines are the two loops unrolled.

s = A2
ss = A1
ss.array_desc_ = nullptr;  // A1.array_desc_ = nullptr;
ss = A3;
ss.array_desc_ = nullptr;  // A3.array_desc_ = nullptr;
s.array_desc_->siblings.clear();  // Clear S2
s = A3;
// segfault when fetching ss cause s.array_desc is nullptr (assuming it didn't)
ss = A1;
ss.array_desc_ = nullptr;  // A1.array_desc_ = nullptr;
ss = A2;
ss.array_desc_ = nullptr;  // A2.array_desc_ = nullptr;
// segfault again when clearing siblings for s
```

Reply:
This works fine:
```
a, b, c = mx.split(mx.zeros(9), 3)
del a, b, c
```

Maybe I'm misunderstanding your comment, but I think it's ok because each array has it's own `array_desc_`?

Reply:
However, there is a bug elsewhere as the C++ tests aren't passing.. so let me track that down. Then I can look into simplifying this function so it's not so confusing.